src/utils/chart.ts

```diff
@@ -0,0 +1,97 @@
+import { AppAction } from '@graasp/sdk';
+
+import {
+  addMinutes,
+  addSeconds,
+  format,
+  formatDuration,
+  getMinutes,
+  intervalToDuration,
+  isBefore,
+  isEqual,
+  parseISO,
+  setMinutes,
+} from 'date-fns';
+
+import { CodeVersionType } from '@/interfaces/codeVersions';
+
+const INTERVAL_DURATION_IN_MINUTES = 10;
+export const VERSION_STEP_DURATION = 2000;
+
+// Function that take starttime and end time and handle intervals between depends on INTERVAL_DURATION_IN_MINUTES
+export const distributeIntervals = ({
+  startDate,
+  endDate,
+}: {
+  startDate: string;
+  endDate: string;
+}): { [key: string]: number } => {
+  const start = parseISO(startDate);
+  const end = addMinutes(endDate, INTERVAL_DURATION_IN_MINUTES);
+  let current = setMinutes(
+    start,
+    Math.floor(getMinutes(start) / INTERVAL_DURATION_IN_MINUTES) *
+      INTERVAL_DURATION_IN_MINUTES,
+  );
+  const group: { [key: string]: number } = {};
+  while (end > current) {
+    group[format(current, "yyyy-MM-dd'T'HH:mm:ssXXX")] = 0;
+    current = addMinutes(current, INTERVAL_DURATION_IN_MINUTES);
+  }
+  return group;
+};
+
+export const groupActionsPerInterval = (
+  intervals: { [key: string]: number },
+  actions: AppAction<CodeVersionType>[],
+): { [key: string]: number } => {
+  const intervalsCopy = { ...intervals };
+  const intervalArr = Object.keys(intervals);
+  intervalArr.forEach((interval, index) => {
+    const totalActionsPerInterval = actions.reduce((acc, action) => {
```

Comment:
This means you iterate `interval x actions` in all cases. Would it be more efficient to iterate over the actions and to push them to the right interval ?